File: src/com/tao/chinachuclient/ProgramListAdapter.java
package com.tao.chinachuclient;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

import Chinachu4j.Program;
import Chinachu4j.Recorded;
import Chinachu4j.Reserve;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.graphics.Color;
import android.graphics.Paint;
import android.preference.PreferenceManager;
import android.text.TextPaint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

public class ProgramListAdapter extends ArrayAdapter<Object>{
	private LayoutInflater mInflater;
	private boolean oldCategoryColor;
	private int type;

	public ProgramListAdapter(Context context, int type){
		super(context, android.R.layout.simple_list_item_1);
		mInflater = (LayoutInflater)context.getSystemService(Activity.LAYOUT_INFLATER_SERVICE);
		oldCategoryColor = PreferenceManager.getDefaultSharedPreferences(context).getBoolean("oldCategoryColor", false);
		this.type = type;
	}

	class ViewHolder{
		TextView title, date;
	}

	@SuppressLint("InflateParams")
	@Override
	public View getView(int position, View convertView, ViewGroup parent){
		final ViewHolder holder;
		Program item;
		if(type == Type.RESERVES)
			item = ((Reserve)getItem(position)).getProgram();
		else if(type == Type.RECORDED)
			item = ((Recorded)getItem(position)).getProgram();
		else
			item = (Program)getItem(position);

		if(convertView == null) {
			convertView = mInflater.inflate(R.layout.program_list_layout, null);
			TextView title = (TextView)convertView.findViewById(R.id.program_title);
			TextView date = (TextView)convertView.findViewById(R.id.program_date);

			holder = new ViewHolder();
			holder.title = title;
			holder.date = date;

			convertView.setTag(holder);
		}else{
			holder = (ViewHolder)convertView.getTag();
		}

		String category = item.getCategory();
		if(oldCategoryColor){
			switch(category){
			case "anime":
				convertView.setBackgroundResource(R.drawable.old_anime);
				break;
			case "information":
				convertView.setBackgroundResource(R.drawable.old_information);
				break;
			case "news":
				convertView.setBackgroundResource(R.drawable.old_news);
				break;
			case "sports":
				convertView.setBackgroundResource(R.drawable.old_sports);
				break;
			case "variety":
				convertView.setBackgroundResource(R.drawable.old_variety);
				break;
			case "drama":
				convertView.setBackgroundResource(R.drawable.old_drama);
				break;
			case "music":
				convertView.setBackgroundResource(R.drawable.old_music);
				break;
			case "cinema":
				convertView.setBackgroundResource(R.drawable.old_cinema);
				break;
			case "etc":
				convertView.setBackgroundResource(R.drawable.old_etc);
				break;
			}
		}else{
			switch(category){
			case "anime":
				convertView.setBackgroundResource(R.drawable.anime);
				break;
			case "information":
				convertView.setBackgroundResource(R.drawable.information);
				break;
			case "news":
				convertView.setBackgroundResource(R.drawable.news);
				break;
			case "sports":
				convertView.setBackgroundResource(R.drawable.sports);
				break;
			case "variety":
				convertView.setBackgroundResource(R.drawable.variety);
				break;
			case "drama":
				convertView.setBackgroundResource(R.drawable.drama);
				break;
			case "music":
				convertView.setBackgroundResource(R.drawable.music);
				break;
			case "cinema":
				convertView.setBackgroundResource(R.drawable.cinema);
				break;
			case "etc":
				convertView.setBackgroundResource(R.drawable.etc);
				break;
			}
		}

		holder.title.setText(item.getTitle());

		String start = new SimpleDateFormat("MM/dd (E) HH:mm", Locale.JAPANESE).format(new Date(item.getStart()));
		String end = new SimpleDateFormat("MM/dd (E) HH:mm", Locale.JAPANESE).format(new Date(item.getEnd()));

		String startDay = start.substring(0, 9);
		String endDay = end.substring(0, 9);

		if(startDay.equals(endDay))
			end = end.substring(10);

		holder.date.setText(start + " 〜 " + end);

		if(type == Type.RESERVES){
			Reserve reserve = (Reserve)getItem(position);
			if(!reserve.getIsManualReserved() && reserve.getIsSkip()){
				holder.title.setTextColor(Color.GRAY);
				holder.date.setTextColor(Color.GRAY);
				TextPaint titlePaint = holder.title.getPaint();
				TextPaint datePaint = holder.date.getPaint();
				titlePaint.setFlags(holder.title.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
				titlePaint.setAntiAlias(true);
				datePaint.setFlags(holder.date.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
				datePaint.setAntiAlias(true);
			}
		}
		return convertView;
	}
}

File: app/src/main/java/com/tao/chinachuclient/ProgramListAdapter.kt
package com.tao.chinachuclient

import android.content.Context
import android.graphics.Color
import android.graphics.Paint.STRIKE_THRU_TEXT_FLAG
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.core.content.ContextCompat
import com.tao.chinachuclient.databinding.ProgramListLayoutBinding
import sugtao4423.library.chinachu4j.Program
import sugtao4423.library.chinachu4j.Recorded
import sugtao4423.library.chinachu4j.Reserve
import java.text.SimpleDateFormat
import java.util.*

class ProgramListAdapter(context: Context, val type: Int) :
    ArrayAdapter<Any>(context, android.R.layout.simple_list_item_1) {

    private val oldCategoryColor =
        (context.applicationContext as App).currentServer.oldCategoryColor
    private val nowYear = Calendar.getInstance().get(Calendar.YEAR).toString()

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val binding = if (convertView == null) {
            val inflater = LayoutInflater.from(context)
            val tBinding = ProgramListLayoutBinding.inflate(inflater, parent, false)
            tBinding.root.tag = tBinding
            tBinding
        } else {
            convertView.tag as ProgramListLayoutBinding
        }

        var item = getItem(position)
        item = when (type) {
            Type.RESERVES -> (item as Reserve).program
            Type.RECORDED -> (item as Recorded).program
            else -> item as Program
        }

        binding.root.setBackgroundResource(
            if (oldCategoryColor) {
                when (item.category) {
                    "anime" -> R.drawable.old_anime
                    "information" -> R.drawable.old_information
                    "news" -> R.drawable.old_news
                    "sports" -> R.drawable.old_sports
                    "variety" -> R.drawable.old_variety
                    "drama" -> R.drawable.old_drama
                    "music" -> R.drawable.old_music
                    "cinema" -> R.drawable.old_cinema
                    "etc" -> R.drawable.old_etc
                    else -> R.drawable.old_etc
                }
            } else {
                when (item.category) {
                    "anime" -> R.drawable.anime
                    "information" -> R.drawable.information
                    "news" -> R.drawable.news
                    "sports" -> R.drawable.sports
                    "variety" -> R.drawable.variety
                    "drama" -> R.drawable.drama
                    "music" -> R.drawable.music
                    "cinema" -> R.drawable.cinema
                    "etc" -> R.drawable.etc
                    else -> R.drawable.etc
                }
            }
        )

        binding.programTitle.apply {
            text = item.title
            paint.isAntiAlias = true
        }
        binding.programDate.apply {
            text = getDateText(item)
            paint.isAntiAlias = true
        }

        if (type == Type.RESERVES) {
            val reserve = getItem(position) as Reserve
            if (!reserve.isManualReserved && reserve.isSkip) {
                binding.programTitle.let {
                    it.setTextColor(Color.GRAY)
                    it.paint.flags = it.paintFlags or STRIKE_THRU_TEXT_FLAG
                }
                binding.programDate.let {
                    it.setTextColor(Color.GRAY)
                    it.paint.flags = it.paintFlags or STRIKE_THRU_TEXT_FLAG
                }
            } else {
                binding.programTitle.let {
                    it.setTextColor(ContextCompat.getColor(context, R.color.titleText))
                    it.paint.flags = it.paintFlags and STRIKE_THRU_TEXT_FLAG.inv()
                }
                binding.programDate.let {
                    it.setTextColor(ContextCompat.getColor(context, R.color.dateText))
                    it.paint.flags = it.paintFlags and STRIKE_THRU_TEXT_FLAG.inv()
                }
            }
        }
        return binding.root
    }

    private fun getDateText(item: Program): String {
        val dateText = run {
            val dateFormat = SimpleDateFormat("yyyy,MM/dd (E),HH:mm", Locale.JAPANESE)
            val startDates = dateFormat.format(Date(item.start)).split(",")
            val endDates = dateFormat.format(Date(item.end)).split(",")

            val startText = startDates.let {
                when {
                    it[0] != nowYear -> "${it[0]}/${it[1]} ${it[2]}"
                    else -> "${it[1]} ${it[2]}"
                }
            }
            val endText = endDates.let {
                when {
                    it[0] != startDates[0] -> "${it[0]}/${it[1]} ${it[2]}"
                    it[1] == startDates[1] -> it[2]
                    else -> "${it[1]} ${it[2]}"
                }
            }
            "$startText 〜 $endText"
        }

        if (type == Type.RESERVES || type == Type.RECORDING) {
            var deltaSec = ((item.start - System.currentTimeMillis()) / 1000).toInt()
            val suffix: String
            if (deltaSec < 0) {
                deltaSec = -deltaSec
                suffix = "前"
            } else {
                suffix = "後"
            }

            return if (deltaSec < 60) {
                "$dateText [%d秒%s]".format(deltaSec, suffix)
            } else {
                var delta = deltaSec.toFloat() / 60
                if (delta < 60.0f) {
                    "$dateText [%.1f分%s]".format(delta, suffix)
                } else {
                    delta /= 60f
                    if (delta < 24.0f) {
                        "$dateText [%.1f時間%s]".format(delta, suffix)
                    } else {
                        delta /= 24f
                        "$dateText [%.1f日%s]".format(delta, suffix)
                    }
                }
            }
        }

        return dateText
    }

    override fun addAll(objects: Array<*>) {
        objects.map {
            add(it)
        }
    }

}
